Replace JSON parser with RapidJSON

lpereira opened the pull request:
This gets rid of the Casablanca JSON parser and replaces it with RapidJSON, which is used in in-situ mode. This reduces temporary allocations quite significantly; so much so that, in my test, this shaves ~5.5ms from startup time of a `webapp` ASP.NET program.

This is a draft PR mostly to gather comments around portability issues (e.g. string vs. wide strings), and possibly some gotchas that I may have missed when converting the code to use this other library.  The APIs are very similar, so it shouldn't be difficult to compare the diff.

RapidJSON is based off of commit [d87b698](https://github.com/Tencent/rapidjson/commit/d87b698d0fcc10a5f632ecbc80a9cb2a8fa094a5), as there has not been a release in a few years now.  I have not tested it with SIMD support, but since that's a compile-time option, I don't know if it's a good idea to use it (maybe SSE2 on x64, as that's part of the ISA, but we can't enable SSE4.2 for everybody; in any case, I'd have to measure to see if it's worth the trouble).

AaronRobinsonMSFT commented:
@lpereira Can you also provide the binary size comparison for both with Casablanca and with RapidJSON. The startup win is awesome!

/cc @jeffschwMSFT

AaronRobinsonMSFT commented on src/corehost/cli/comhost/clsidmap.cpp:
```diff
@@ -79,9 +75,9 @@ namespace
-            json::object &val = prop.second.as_object();
-            e.assembly = val.at(_X("assembly")).as_string();
-            e.type = val.at(_X("type")).as_string();
+            auto &val = prop.value.GetObject();
```
I am a huge fan of `auto` when the type is observed on the right side _or_ when using iterators. In this case, the type is opaque from usage and difficult to determine in review. I personally am okay with this but other might have a different opinion, just an FYI.

elinor-fung replied:
Personally, I'm a fan of the explicit type.

AaronRobinsonMSFT commented on src/corehost/cli/fxr/hostpolicy_resolver.cpp:
```diff
@@ -27,49 +29,27 @@ namespace
+        json_parser_t json;
+        if (!json.parse_file(deps_json))
         {
-            trace::verbose(_X("Dependency manifest [%s] could not be opened"), deps_json.c_str());
```
What is the reasoning behind removing the logging?

lpereira replied:
json_parser_t::parse_file() will print generic error messages (mentioning file names).  That seems sufficient.

AaronRobinsonMSFT commented on src/corehost/cli/json.h:
```diff
@@ -0,0 +1,29 @@
+        bool parse_string(const pal::string_t& str, const pal::string_t& context);
+
+    private:
+        pal::char_t *m_json{nullptr};
```
~~pal::char_t *m_json{nullptr};~~ -> `std::unique_ptr<pal::char_t[]> m_json;`

AaronRobinsonMSFT replied:
Then the destructor can be removed.

lpereira replied:
> @lpereira Can you also provide the binary size comparison for both with Casablanca and with RapidJSON. The startup win is awesome!
> 

These are the sizes with RapidJSON (release mode, Linux, x86-64, stripped):
```
   text	   data	    bss	    dec	    hex	filename
 306034	   2292	   1056	 309382	  4b886	libhostfxr.so
 258763	   2612	    800	 262175	  4001f	libhostpolicy.so
```

And these are the sizes with Casablanca (release mode, Linux, x86-64, stripped):
```
   text	   data	    bss	    dec	    hex	filename
 340175	   4340	   1088	 345603	  54603	libhostfxr.so
 294478	   4636	    840	 299954	  493b2	libhostpolicy.so

```

Which is pretty nifty, too: ~69KB of text + ~9KB of data shaved, for ~79KB of gain in size.

lpereira replied:
Old habits die hard. Fixed.

lpereira commented:
> @lpereira Can you also provide the binary size comparison for both with Casablanca and with RapidJSON. The startup win is awesome!
> 

These are the sizes with RapidJSON (release mode, Linux, x86-64, stripped):
```
   text	   data	    bss	    dec	    hex	filename
 306034	   2292	   1056	 309382	  4b886	libhostfxr.so
 258763	   2612	    800	 262175	  4001f	libhostpolicy.so
```

And these are the sizes with Casablanca (release mode, Linux, x86-64, stripped):
```
   text	   data	    bss	    dec	    hex	filename
 340175	   4340	   1088	 345603	  54603	libhostfxr.so
 294478	   4636	    840	 299954	  493b2	libhostpolicy.so

```

Which is pretty nifty, too: ~69KB of text + ~9KB of data shaved, for ~79KB of gain in size.

lpereira commented:
Force-pushed to fix a build issue (it was not picking up the copy of RapidJSON in the repository; I have it installed as part of the system so the build was working just fine but broke in CI), and addressed some of the comments by @AaronRobinsonMSFT (thanks!)

elinor-fung commented on src/corehost/cli/json.h:
```diff
@@ -0,0 +1,28 @@
+#include <memory>
+#include "pal.h"
+#include "rapidjson/document.h"
+#include "rapidjson/fwd.h"
```
Maybe I'm missing it, but do the types defined in `fwd.h` always use `char` as the character type instead of the whole platform-dependent `char`/`wchar` thing we do? (I'm not familiar with RapidJSON at all - just searched for one of the types being used in this change)

lpereira replied:
Yes, you're right.  I've declared the types we need inside `json_parser_t` to take into account the different encodings and character type width.

elinor-fung commented on src/corehost/cli/json.h:
```diff
@@ -0,0 +1,28 @@
+#include "rapidjson/document.h"
+#include "rapidjson/fwd.h"
+
+class json_parser_t {
```
Could we name the .h/.cpp file `json_parser` instead of just `json` since this is the only thing in it?

lpereira replied:
Sure.

elinor-fung commented on src/corehost/cli/deps_format.h:
```diff
@@ -116,6 +114,8 @@ class deps_json_t
 
     pal::string_t m_deps_file;
+
+    json_parser_t m_json;
```
Why do we need to hang on to this? Do we have pointers to its buffer after parsing? And can you add a comment about it here?

lpereira replied:
That used to be the case, yes, and ended up forgetting to remove this.  Will be declared in `deps_json_t::load()`.

vitek-karas commented on src/corehost/cli/comhost/clsidmap.cpp:
```diff
@@ -79,9 +75,9 @@ namespace
-            e.assembly = val.at(_X("assembly")).as_string();
-            e.type = val.at(_X("type")).as_string();
+            auto &val = prop.value.GetObject();
+            e.assembly = val[_X("assembly")].GetString();
```
The encoding/chartype will be a bit tricky. We obviously parse the input as UTF8 (as we should), but the way it's currently written it also means that all of the JSON is handled as UTF8. Works great on Linux - not so much on Windows. So we would either have to somehow ask RapidJSON to transcode (not sure if it can do that), or convert the values after the fact. Would be worth measuring which way is better - also would be worth investigating how this worked with the previous parser.
As for lookups - I would not mind if we specify the property names as UTF8 always (even on Windows) - they are pure ASCII anyway and it's smaller that way. But we need the returned values to be UTF16 on Windows (as they can be anything not just pure ASCII).

Because of this, the code doesn't even come close to compiling on Windows.

AaronRobinsonMSFT replied:
>Because of this, the code doesn't even come close to compiling on Windows.

I am not trying to be pedantic here, but there is nothing in this or the previous logic that is dealing with encoding. And the compiler error isn't related to encoding. We are hitting issues with code unit size (i.e. the smallest defined unit of storage for an encoding). My understanding is that the encoding of the JSON files will always be UTF-8. Is the JSON on Windows UTF-16? If the JSON is always UTF-8, we can consume the file using the `char` code unit on all platforms and all will be well. If the JSON is always UTF-16 on Windows, then this is a bit of a mess because that would mean JSON files created on Windows couldn't be used off Windows.

The code unit size on Windows is 16-bits (i.e. `WCHAR`) and encoded as UTF-16. This becomes the underlying problem though because even in Casablanca I don't believe any detection or conversion of encoding was occurring. All that was happening to my knowledge was blind widening of code units. I know I didn't try testing an assembly or type name containing a non-ASCII character.

The spirit of what @vitek-karas is saying is correct, I just want to ensure we are using the correct terminology for this type of issue since it is so easy to make mistakes.

lpereira replied:
I looked up the [JSON RFC](https://tools.ietf.org/html/rfc7159) and here's what I found:

```
8.1.  Character Encoding

   JSON text SHALL be encoded in UTF-8, UTF-16, or UTF-32.  The default
   encoding is UTF-8, and JSON texts that are encoded in UTF-8 are
   interoperable in the sense that they will be read successfully by the
   maximum number of implementations; there are many implementations
   that cannot successfully read texts in other encodings (such as
   UTF-16 and UTF-32).

   Implementations MUST NOT add a byte order mark to the beginning of a
   JSON text.  In the interests of interoperability, implementations
   that parse JSON texts MAY ignore the presence of a byte order mark
   rather than treating it as an error.
```

So I need to look in Casablanca to know how they handle encoding stuff (as @elinor-fung pointed out, `fwd.h` uses `char` on all systems).

I'm curious though about the BOM -- the standard says that it should never be there, but are we generating JSON files with them?

AaronRobinsonMSFT replied:
Unfortunately the RPF specification is an orthogonal issue. The issue here is what will the SDK produce since we aren't officially support arbitrary JSON files.

>I'm curious though about the BOM -- the standard says that it should never be there, but are we generating JSON files with them?

Be careful not to confuse a standard with an RFC. Anyone can create an RFC but they are non-binding definitions, not standards (i.e. ISO or ECMA). The BOM is only one place that is an issue. At least one of the Microsoft supplied JSON serializers will also add a trailing `,` on the last field for an object which is also not permitted per the JSON RFC.

My point here is we should adhere to the contract that the SDK provides for all JSON assets and confirm the encoding is consistent and probably add a test for non-ASCII characters in managed COM server names.

AaronRobinsonMSFT replied:
Looks like the `.clsidmap` file will always be UTF-8 - https://github.com/dotnet/sdk/blob/1f061a6e33480947ba81791691ff74f413200a62/src/Tasks/Microsoft.NET.Build.Tasks/ComHost.cs#L43-L44

However, I don't see the encoding specified for the `.deps.json` - https://github.com/dotnet/sdk/blob/1f061a6e33480947ba81791691ff74f413200a62/src/Tasks/Microsoft.NET.Build.Tasks/GenerateDepsFile.cs#L171-L174 or `.runtimeconfig.json` - https://github.com/dotnet/sdk/blob/1f061a6e33480947ba81791691ff74f413200a62/src/Tasks/Microsoft.NET.Build.Tasks/GenerateRuntimeConfigurationFiles.cs#L312-L320.

AaronRobinsonMSFT replied:
Based on a cursory search for `DependencyContextWriter` (used for `.deps.json`) it seems this could be defined to use UTF-8 encoding, but there appears to be some definitions around that rely on the default. The `.runtimeconfig.json` appears to be using defaults in all cases so that would need to be determined as well.

lpereira replied:
The function to skip the BOM is named `skip_utf8_bom()`, which hints that this is indeed always UTF-8. The UTF-16 BOM uses a different byte sequence (0xff,0xfe or 0xfe,0xff depending on the endianess, as opposed to the [fixed] 0xef, 0xbb, 0xbf from UTF8).

vitek-karas replied:
I think we should look at this from backward compat perspective - try to figure out what the old parser did... and either emulate that or knowingly change it.
I'm pretty sure that SDK will always produce UTF8 and so on (baring the BOM discussions which I would have to investigate in detail) - so we need that to work, but in reality we want to at the very least know about what worked before and what will work now.

Since this is obviously a "Tricky" topic - I think we should add tests for these cases as well (for all our JSON inputs ideally, but at the very least for `.runtimeconfig.json` and `.deps.json`).

lpereira replied:
Looked briefly at the old parser, and they have the equivalent of `pal::string_t` (`utility::string_t`), that's defined to either `std::wstring` or `std::string` if `_UTF16_STRINGS` is defined (which is always true if `_WIN32` is also defined). The `as_string()` method of a JSON value in the old parser just returns the internal `utility::string_t` value.

lpereira replied:
OK, this was easier than I thought.  Unfortunately we won't be able to use in-situ parsing on Windows, because UTF-8 data will need to be converted to UTF-16 data before parsing (which is luckily easy to do in RapidJSON).  I don't have numbers to determine the impact, though, but gut instinct says it won't be worse than Casablanca (at least that's what it did in a nutshell).

lpereira replied:
I'll wait for the CI to tell if I need to do anything else; when I get the answer, I'll see about writing the tests.  Might be a good idea to send another PR too to force data to be always written in UTF-8 and not rely on any defaults.

lpereira replied:
Seems to work fine now. Tests are passing on Windows too.

vitek-karas commented:
We'll also need to add attribution to the https://github.com/dotnet/core-setup/blob/master/THIRD-PARTY-NOTICES.TXT (or maybe elsewhere - I'm not sure exactly how this works)

lpereira commented:
> We'll also need to add attribution to the https://github.com/dotnet/core-setup/blob/master/THIRD-PARTY-NOTICES.TXT (or maybe elsewhere - I'm not sure exactly how this works)

Absolutely -- I'll also add a README file pointing to RapidJSON, with information about which commit hash I obtained the files from, this kind of stuff.

lpereira commented:
(Force-pushed with some changes, addressing some of the comments here -- but mostly to check if this will actually build on Windows.)

AaronRobinsonMSFT commented on src/corehost/cli/comhost/clsidmap.cpp:
```diff
@@ -45,29 +45,21 @@ namespace
-            json_map = json::value::parse(json_map_raw);
-        }
-        catch (const json::json_exception&)
+        if (!json.parse_stream(json_map_raw, "<embedded .clsidmap>"))
```
I think you are going to need `_X("<embedded .clsidmap>")`

lpereira replied:
Yes, indeed. Fixed.

AaronRobinsonMSFT commented on src/corehost/cli/json_parser.cpp:
```diff
@@ -0,0 +1,94 @@
+bool json_parser_t::parse_string(const pal::string_t& str, const pal::string_t& context)
+{
+    realloc_buffer(str.length());
+    memcpy(m_json.get(), str.c_str(), str.length());
```
The code unit size should be considered whenever using low level C memory functions like `memcpy`.

``` c++
size_t byte_len = str.length() * sizeof(str[0]);
realloc_buffer(byte_len);
memcpy(m_json.get(), str.c_str(), byte_len);
```

lpereira replied:
Argh. Yeah, adding this to the bag of stuff to worry about. Not really used to wide strings.

lpereira replied:
Pushed a new version where this change won't b e necessary; data in m_json will be kept in UTF-8, and m_document will handle the UTF8 -> UTF-16 goo.

AaronRobinsonMSFT commented on src/corehost/cli/json_parser.h:
```diff
@@ -0,0 +1,36 @@
+    private:
+        // JSON files parsed by this class are always encoded in UTF-8, so use char[]
+        // instead of pal::char_t[] for m_json.
+        std::unique_ptr<char[]> m_json;
```
I didn't realize what you were doing here when I suggested this. In this case, using `std::vector<char>` might be easier since it provides many resize and manipulation functions.

lpereira replied:
Done.

lpereira commented:
Force-pushed after cleaning up the encoding story: data is read from the file as UTF-8, but processed as UTF-16 on Windows.  This unfortunately means that we can't do in-situ parsing on Windows at this point; we might be able to convert the UTF-8 data to UTF-16 and parse that in-situ, but I'll probably do that later.  At this point I'm only looking towards making this work on Windows too, things can be improved later.

lpereira commented:
@elinor-fung wrote:

 > How hard would it be to also get time/size diffs on Windows? smiley

Now that this is building on Windows, it should be possible to get the size diff with ease.  I imagine performance will be slightly better, but not as good as on non-Windows platforms due to in-situ parsing not being used, though.

lpereira commented:
The test failures are expected: they're testing malformed JSON files, and output has changed.  Will need to change the tests themselves.

lpereira commented:
Finally all tests are passing on Windows.

vitek-karas commented on src/test/HostActivation.Tests/CommandExtensions.cs:
```diff
@@ -36,6 +36,11 @@ public static Command EnableTracingAndCaptureOutputs(this Command command)
                 .CaptureStdErr();
         }
 
+        public static Command WithoutDotNetRoot(this Command command)
```
Nit: I would prefer "positive" methods - in this case I think it would make sense to be able to simply call `.DotNetRoot(null)` (to unset the variable).

lpereira commented on src/test/HostActivation.Tests/DependencyResolution/ResolveComponentDependencies.cs:
```diff
@@ -170,7 +170,7 @@ public void ComponentWithCorruptedDepsJsonShouldFail()
                 .And.HaveStdOutContaining($"corehost_resolve_component_dependencies:Fail[0x{Constants.ErrorCode.ResolverInitFailure.ToString("x")}]")
                 .And.HaveStdOutContaining("corehost reported errors:")
-                .And.HaveStdOutContaining($"A JSON parsing exception occurred in [{component.DepsJson}]: * Line 1, Column 2 Syntax error: Malformed token")
+                .And.HaveStdOutContaining($"A JSON parsing exception occurred in [{component.DepsJson}]: Invalid value.")
```
I don't know how to get a better error message from RapidJSON; maybe I can get the line/column, but the error code in many situations is just that. I'll try getting the file location as that's useful information.

vitek-karas commented on src/corehost/cli/json_parser.cpp:
```diff
@@ -0,0 +1,136 @@
+        {
+            // If this condition is reached, offset is at least 1.  So reading *(ptr+1)
+            // should be fine, as there should be at least two characters from ptr
+            // onwards.
```
I don't see how offset being at least 1 affects the size of the `ptr` buffer. Can we please pass the `json` as the vector and thus get automatic range checks? Perf is not a big deal in this error case.

vitek-karas commented on src/corehost/cli/json_parser.cpp:
```diff
@@ -0,0 +1,130 @@
+{
+    *line = *column = 1;
+
+    for (size_t i = 0; i < offset; i++)
```
Nit: can you add an assert that offset fits into the vector? Just paranoia - and also acts as a documentation kind of thing.

AaronRobinsonMSFT replied:
>and also acts as a documentation kind of thing.

I do this on a regular basis. Asserts are easy to change and I find that they are great for giving context to future developers about how I was thinking about a feature. The assert doesn't necessarily mean there is an issue, rather it can mean "I only verified the following and anything more should probably be thought about and/or tested". /cc @elinor-fung.

elinor-fung commented on src/corehost/cli/json_parser.cpp:
```diff
@@ -0,0 +1,130 @@
+        int line, column;
+        size_t offset = m_document.GetErrorOffset();
+
+        get_line_column_from_offset(m_json, offset, &line, &column);
```
I was going to ask if we wanted to file something in rapidjson suggesting supporting this and found they already had [916](https://github.com/Tencent/rapidjson/issues/916). Is the CursorStreamWrapper that was added something we could use to figure out the line/column?

lpereira replied:
Yeah, I found the same issue in their repo; that's why I wrote the function.  I haven't looked at CursorStreamWrapper yet; will see if it works too.

lpereira replied:
From a first glance, it doesn't seem I can use in-situ parsing when using CursorStreamWrapper; it wraps an InputStream whereas the in-situ parsing mode expects a character array compatible with the input encoding type.

Even if that's not a problem, since errors should be rare occurrences, I'd rather not add the (tiny) overhead of counting lines and columns except if absolutely necessary.

Also, it doesn't handle `\r\n` line endings, only `\n`.  (Not a big deal, as columns would be off by 1, which would be good enough in most cases to determine the location of an error, but still...)

lpereira commented on src/test/HostActivation.Tests/CommandExtensions.cs:
```diff
@@ -38,6 +38,11 @@ public static Command EnableTracingAndCaptureOutputs(this Command command)
 
         public static Command DotNetRoot(this Command command, string dotNetRoot)
         {
+            if (dotNetRoot == null)
```
Good idea.

lpereira commented:
Force-pushed addressing comments.

lpereira commented:
Yeah, I'm rebooting into Windows in a few minutes to gather this information.

lpereira commented:
> I'd still be interested in seeing the Windows sizes / times too.

If I did everything right, here are the sizes (used `size` from mingw in WSL). Same deal, Windows, x86-64, release, stripped binaries.  Before:

```
   text    data     bss     dec     hex filename
546434   14504       0  560938   88f2a hostpolicy.dll
554414   11952       0  566366   8a45e hostfxr.dll
```

After:

```
   text    data     bss     dec     hex filename
530634   10928       0  541562   8437a hostfxr.dll
523864   13480       0  537344   83300 hostpolicy.dll
```

So there's some size reduction on Windows as well, although not quite as dramatic (~48KB overall).  The timings I got seem strange; I'll reboot and try again.
